tensorrt_llm/_torch/pyexecutor/scheduler.py

```diff
@@ -216,3 +216,54 @@ def schedule_request(self, active_requests: RequestList,
                                list(generation_requests), list(paused_requests),
                                list(fitting_disagg_gen_init_requests),
                                len(fitting_requests))
+
+
+class ContextBatchingScheduler(SimpleScheduler):
+    """ContextBatchingScheduler batches context requests when scheduling.
+
+    Tries to batch context requests up to max_pending_requests.
+    Schedule immediately, if there are no generation request,
+    or if no new context requests arrive for max_pending_iterations.
+
+    Can be enabled via SchedulerConfig.
+    """
+
+    def __init__(self, capacity_scheduler: CapacityScheduler,
+                 micro_batch_scheduler: MicroBatchScheduler,
+                 max_pending_requests: int, max_pending_iterations: int):
+        super().__init__(capacity_scheduler, micro_batch_scheduler)
+        self.max_pending_requests = max_pending_requests
+        self.max_pending_iterations = max_pending_iterations
+
+        self.last_pending_requests = 0
+        self.pending_iterations = 0
+
+    def schedule_request(self, active_requests: RequestList,
+                         inflight_request_ids: set[int]) -> SchedulerOutput:
+        output = super().schedule_request(active_requests, inflight_request_ids)
+
+        num_pending_requests = len(output.context_requests)
+        if num_pending_requests == 0 or num_pending_requests >= self.max_pending_requests or len(
+                output.generation_requests) == 0:
+            # If we get enough context requests, or we have only context requests, process them
+            self.last_pending_requests = 0
+            self.pending_iterations = 0
+        else:
```

Comment:
Would remove the `else` and return output early.

```suggestion
            self.pending_iterations = 0
            return output
```